### DISCUSSION TITLE
Language Support

### DISCUSSION THREAD

**Discussion by KrishavRajSingh** [cvat-ai/cvat#7975] [Ideas]
Hey, I was trying to work on #374 

We can easily install i18n in the cvat-ui and configure the components to get the translation file via hooks and props. We have 2 option, either keep the translation file in the frontend or we can make a api call to get the translation file.

Making a api call is preferred as loading all the translation might make the front-end heavy and will take time to load.

I have tried making  a endpoint at 'localhost:7000/locales/{lng}/translation.json' and I am able to get the data at this endpoint but for some reasons I am not able to get this at 'localhost:3000/locales/{lng}/translation.json'. I thought it might work just like 'localhost:3000/api/...' . Also I am not able to call 'localhost:7000/locales/{lng}/translation.json' directly from the cvat-ui.
![image](https://github.com/cvat-ai/cvat/assets/99531396/7c472107-9e43-4305-a573-218299954068)

Can you suggest me how to fix this.

And if you have any other alternatives please let me know.
Thanks.

**Comment by acecode** [comment]
I'm using [i18next-resources-to-backend + webpack](https://github.com/i18next/i18next-resources-to-backend?tab=readme-ov-file#you-can-also-lazy-load-the-in-memory-translations-ie-when-using-webpack)
this way webpack will pack resouce files

For this error says `URL scheme`，maybe you should add `http://` or `//` at begin of the url

    `//` means use the same with current page URL

There will be another problems
1.  the static site is start at localhost:3000 and the API server is localhost:7000
    * i18n resources is more related to Static(frontend) server than the API(backend) Server
    * I suggest put locales in `cvat-ui` project

2. `src/locales` folder is not served on url `localhost:3000/locales/`
    * `src/locales` folder is not copy to `dist` by default in dev/prod env

3.  Use an online service like [locize.com](http://locize.com) give us the benefit to manage the resouces and for other ones to 
contribute. The cons is not support no net connection.

**Reply by KrishavRajSingh** [reply]
The main issue I am facing now is the endpoint is at localhost:7000/locales and I am not able to call ot from cvat-ui.

I can only call localhost:3000.

**Reply by acecode** [reply]
Webpack dev config only proxy some path to api server
[webpack.config.js DevServer Proxy Path](https://github.com/cvat-ai/cvat/blob/729322f40e16c28c329021ebd76680326dbeef6a/cvat-ui/webpack.config.js#L70)

```js
/\/api\/.*|analytics\/.*|static\/.*|admin(?:\/(.*))?.*|profiler(?:\/(.*))?.*|documentation\/.*|django-rq(?:\/(.*))?/gm
```

Default cvat-ui can't call ajax api on different domain

`localhost:3000` -> `localhost:7000`

I suggest again that keep the `locales resource` in `cvat-ui` asides the frontend code

    ```js
    // component/x.ts
    ...
    t('Save')
    ...
    
    // locales/zh-cn.json
    ...
      Save: "保存"
    ...
    ``` 

Also We could share some common/basic resource with the `api server`

**Reply by KrishavRajSingh** [reply]
I made an api call to http://localhost:3000/api/locales/..., it is working fine. You can checkout this [PR](https://github.com/KrishavRajSingh/cvat/tree/feat/LanguageSupport).
Thanks a lot for your suggestion. Always open to reviews!!!

> Webpack dev config only proxy some path to api server [webpack.config.js DevServer Proxy Path](https://github.com/cvat-ai/cvat/blob/729322f40e16c28c329021ebd76680326dbeef6a/cvat-ui/webpack.config.js#L70)
> 
> ```js
> /\/api\/.*|analytics\/.*|static\/.*|admin(?:\/(.*))?.*|profiler(?:\/(.*))?.*|documentation\/.*|django-rq(?:\/(.*))?/gm
> ```
> 
> Default cvat-ui can't call ajax api on different domain
> 
> `localhost:3000` -> `localhost:7000`
> 
> I suggest again that keep the `locales resource` in `cvat-ui` asides the frontend code
> 
> ```
> ```js
> // component/x.ts
> ...
> t('Save')
> ...
> 
> // locales/zh-cn.json
> ...
>   Save: "保存"
> ...
> ``` 
> ```
> 
> Also We could share some common/basic resource with the `api server`

**Comment by acecode** [comment]
1. Maybe we should remove the host and port in [i18n.ts loadPath ](https://github.com/KrishavRajSingh/cvat/blob/08bc660cc4e873df5bf1a57f48c56861c9880868/cvat-ui/src/i18n.ts#L27)
```diff
backend: {
-            loadPath: 'http://localhost:3000/api/locales/{{lng}}/{{ns}}.json', // Adjust this path based on your project structure
+            loadPath: '/api/locales/{{lng}}/{{ns}}.json', // Adjust this path based on your project structure
},
```
this host and port may not work in production(docker) env

**Comment by Twenty3hree** [comment]
> Hey, I was trying to work on #374
> 
> We can easily install i18n in the cvat-ui and configure the components to get the translation file via hooks and props. We have 2 option, either keep the translation file in the frontend or we can make a api call to get the translation file.
> 
> Making a api call is preferred as loading all the translation might make the front-end heavy and will take time to load.
> 
> I have tried making a endpoint at 'localhost:7000/locales/{lng}/translation.json' and I am able to get the data at this endpoint but for some reasons I am not able to get this at 'localhost:3000/locales/{lng}/translation.json'. I thought it might work just like 'localhost:3000/api/...' . Also I am not able to call 'localhost:7000/locales/{lng}/translation.json' directly from the cvat-ui. ![image](https://private-user-images.githubusercontent.com/99531396/335783782-7c472107-9e43-4305-a573-218299954068.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE1NTI1NTYsIm5iZiI6MTczMTU1MjI1NiwicGF0aCI6Ii85OTUzMTM5Ni8zMzU3ODM3ODItN2M0NzIxMDctOWU0My00MzA1LWE1NzMtMjE4Mjk5OTU0MDY4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTE0VDAyNDQxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFkZjMzZDc5ZGY3MDEzZmQyNWY3M2Q3ODVlYmQzMjA3OTM5ZTMzYzA3NGNlYmRkOTRkY2FjZTBmNDJiMjMyNjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.T46d1qV1NC9xRXbau3jaUOjipp1GOV-3QnhZ_yk8Dtc)
> 
> Can you suggest me how to fix this.
> 
> And if you have any other alternatives please let me know. Thanks.

So, may I ask about, i18n is added up to main branch？or is it in consideration
